docs/developer-docs/backend/rust/12-searching-records.md

```diff
@@ -0,0 +1,390 @@
+-   How to write a simple search function with partial string matching. 
+-   How profiles are associated with a specific principal.
+
+## Prerequisites
+
+Before getting started, assure you have set up your developer environment according to the instructions in the [developer environment guide](./3-dev-env.md).
+
+## Creating a new project
+
+Open a terminal window on your local computer, if you don’t already have one open. Then, create a new project by running the following command:
+
+``` bash
+dfx new --type=rust rust_profile
+cd rust_profile
+```
+
+Now that you have the files in place for your Rust dapp, we can replace the template `lib.rs` dapp with the Rust dapp we want to deploy on the Internet Computer blockchain.
+
+To replace the default program, open the `src/rust_profile_backend/Cargo.toml` file in a text editor and add `serde` to dependencies.
+
+``` toml
+[dependencies]
+candid = "0.8.2"
+ic-cdk = "0.6.0"
+ic-cdk-macros = "0.6.0"
+serde = "1.0"
+```
+
+Then, open the template `src/rust_profile_backend/src/lib.rs` file in a text editor and delete the existing content.
+
+The next step is to add a Rust program that implements the `getSelf`, `update`, `get`, and `search` functions. To do this, copy and paste this code into the `src/rust_profile_backend/src/lib.rs` file.
+
+```rust
+use ic_cdk::{
+    api::call::ManualReply,
+    export::{
+        candid::{CandidType, Deserialize},
+        Principal,
+    },
+};
+use ic_cdk_macros::*;
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+
+type IdStore = BTreeMap<String, Principal>;
+type ProfileStore = BTreeMap<Principal, Profile>;
+
+#[derive(Clone, Debug, Default, CandidType, Deserialize)]
+struct Profile {
+    pub name: String,
+    pub description: String,
+    pub keywords: Vec<String>,
+}
+
+thread_local! {
+    static PROFILE_STORE: RefCell<ProfileStore> = RefCell::default();
+    static ID_STORE: RefCell<IdStore> = RefCell::default();
+}
+
+#[query(name = "getSelf")]
+fn get_self() -> Profile {
+    let id = ic_cdk::api::caller();
+    PROFILE_STORE.with(|profile_store| {
+        profile_store
+            .borrow()
+            .get(&id)
+            .cloned().unwrap_or_default()
+    })
+}
+
+#[query]
+fn get(name: String) -> Profile {
+    ID_STORE.with(|id_store| {
+        PROFILE_STORE.with(|profile_store| {
+            id_store
+                .borrow()
+                .get(&name)
+                .and_then(|id| profile_store.borrow().get(id).cloned()).unwrap_or_default()
+        })
+    })
+}
+
+#[update]
+fn update(profile: Profile) {
+    let principal_id = ic_cdk::api::caller();
+    ID_STORE.with(|id_store| {
+        id_store
+            .borrow_mut()
+            .insert(profile.name.clone(), principal_id);
+    });
+    PROFILE_STORE.with(|profile_store| {
+        profile_store.borrow_mut().insert(principal_id, profile);
+    });
+}
+
+#[query(manual_reply = true)]
```

Comment:
We should be consistent about either cloning or using ManualReply, not both in similar contexts. Maybe we could start with cloning in both cases, and then include a section afterwards that discusses the purpose of ManualReply and adapts the example to use it.